ballot/service/src/main.rs

```diff
@@ -29,16 +29,16 @@ impl Ballot {
         }
     }
 
-    /// Returns the candidates being voted upon.
-    pub fn candidates(&self, _ctx: &Context) -> Vec<&str> {
-        self.candidates.iter().map(String::as_ref).collect()
-    }
-
     /// Returns the description of this ballot.
     pub fn description(&self, _ctx: &Context) -> &str {
         &self.description
     }
 
+    /// Returns the candidates being voted upon.
+    pub fn candidates(&self, _ctx: &Context) -> Vec<&str> {
+        self.candidates.iter().map(String::as_ref).collect()
```

Comment:
why not return `&[String]`? should be same

Reply:
I think the benefit of the current definition is the parallel to `description()` - one returns a `&str` with the description, and the other returns a `Vec<&str>` with a `&str` for each candidate. Assuming most people doing this tutorial are relatively new to Rust, I think this visual similarity reduces some complexity, if the functionality is the same

Reply:
```rust
pub fn candidates(&self, _ctx: &Context) -> &[String] {
    &self.candidates
}
```

it's simple and efficient--no copying, no allocation